pynamodb/asyncio/batch_get.py

```diff
@@ -0,0 +1,62 @@
+import asyncio
+import typing
+
+from typing import List, Dict, Any, TypeVar
+
+from aioitertools.asyncio import as_completed
+from more_itertools import chunked
+
+from pynamodb.constants import BATCH_GET_PAGE_LIMIT
+
+if typing.TYPE_CHECKING:
+    from pynamodb.models import Model # noqa: F401
+
+_T = TypeVar("_T", bound="Model")
+
+class BatchGetIterator(typing.AsyncIterator[_T]):
+
+    def __init__(self, model_cls: typing.Type[_T], keys: typing.Iterable, consistent_read: bool | None = None, attributes_to_get: typing.Sequence[str] | None = None):
+        self.model_cls = model_cls
+        self.unprocessed_batch_items = list(keys)
+        self.consistent_read = consistent_read
+        self.attributes_to_get = attributes_to_get
+        self.current_batch: List[Dict[str, Any]] = []
+
+    def __aiter__(self):
+        return self
+
+    def _get_coroutines(self, unprocessed_batch_items) -> list[typing.Awaitable]:
+        """Split unprocessed batch items into chunks and create coroutines for each chunk."""
+        serialized_keys = list(self.model_cls._batch_serialize_keys(unprocessed_batch_items))
+        return [
+            self.model_cls._async_batch_get_item(
+                chunk, self.consistent_read, self.attributes_to_get
+            )
+            for chunk in chunked(serialized_keys, BATCH_GET_PAGE_LIMIT)
+        ]
+
+    @property
+    def timeout(self) -> int:
+        return self.model_cls.Meta.connect_timeout_seconds + self.model_cls.Meta.read_timeout_seconds
+
+    async def __anext__(self) -> _T:
+        # First check if we have items awaiting to be processed
+        if self.current_batch:
+            item = self.current_batch.pop(0)
+            await asyncio.sleep(0)
+            return self.model_cls.from_raw_data(item)
+
+        # Check if we have unprocessed items
+        if self.unprocessed_batch_items:
+            # Process next batch
+            coros = self._get_coroutines(self.unprocessed_batch_items)
+            self.unprocessed_batch_items = []
+            async for items, unprocessed_keys in as_completed(coros, timeout=self.timeout):
```

Comment:
This may be better with using `asyncio.FIRST_COMPLETED`, so that the `__anext__` does not wait for everything on the first call?

Something like

```python
done, pending = asyncio.wait(coros, timeout=self.timeout, return_when=FIRST_COMPLETED)
self.unprocessed_batch_items.extend(pending)
self.current_batch.extend(done)
...
```

Would be more in the spirit of a generator, I feel

Reply:
I went through some some different designs for this piece and `wait` was on the list but I had two problems with it:
1. A task being pending is not the same as a task being complete but having an unprocessed amount of items (from AWS). Which would mean keeping track of both every iteration, which was fine but much more verbose than I liked.
2. To avoid structural concurrency problems with tasks being ready but not collected I preferred to process them all in a single call instead of collecting them through the iterations.

That's definitely something that I want to re-visit but only applies to really large `batch_get` operations which we can optimize when we get there